### PR TITLE
Add AWS Cloudtrail event source plugin

### PR DESCRIPTION
**Pull request by abikouo**
This PR adds support to AWS Cloudtrail as an event source and includes an example rulebook.

### CODE REVIEW

File: plugins/event_source/aws_cloudtrail.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+"""
+aws_cloudtrail.py
+
+An ansible-rulebook event source module for getting events from an AWS CloudTrail
+
+Arguments:
+    connection - Parameters use to create AWS session
+    lookup_attributes - List of attributes to filter the events returned. each item of the list should contain an attribute key and value.
```
Comment (jillr) [inline plugins/event_source/aws_cloudtrail.py:8]:
```suggestion
    connection - Parameters used to create AWS session
    lookup_attributes - List of attributes to filter the events returned. Each item of the list should contain an attribute key and value.
```

File: plugins/event_source/aws_cloudtrail.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    - ansible.eda.aws_cloudtrail:
+        connection:
+            aws_access_key_id: access123456789
+            aws_secret_access_key: secret123456789
+            region_name: us-east-1
+            profile_name: default
+        lookup_attributes:
+            - key: eventSource
+              value: ec2.amazonaws.com
```
Comment (jillr) [inline plugins/event_source/aws_cloudtrail.py:22]: What do you think about having `lookup_attributes` take a mapping of key:value pairs and convert them in the plugin to the format that boto requires?  Similar to `amazon.aws.module_utils.transofrmation.ansible_dict_to_boto3_filter_list` or  `amazon.aws.module_utils.tagging.ansible_dict_to_boto3_tag_list`, like this:

```
        lookup_attributes:
            eventSource: ec2.amazonaws.com
```

Comment (abikouo) [reply plugins/event_source/aws_cloudtrail.py:22]: Good idea, I will do that

**Review by benthomasson — DISMISSED** [review]

**Review by jillr — APPROVED** [review]
LGTM after a rebase

**Comment by abikouo** [comment]
> LGTM after a rebase

@jillr pull request has been rebased
Thanks

**Review by jillr — APPROVED** [review]

File: plugins/event_source/aws_cloudtrail.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            event_ids = [e["EventId"]]
+        elif event_time == e["EventTime"]:
+            event_ids.append(e["EventId"])
+        result.append(e)
+    return result, event_time, event_ids
+
+
+class AnsibleAwsCloudTrailEventClient(object):
+    def __init__(self, connection):
+        session = boto3.session.Session(**connection)
```
Comment (bzwei) [inline plugins/event_source/aws_cloudtrail.py:80]: Seems you should use aiobotocore.session. Take a look at similar implementation at https://github.com/ansible/event-driven-ansible/blob/main/plugins/event_source/sqs.py

File: plugins/event_source/aws_cloudtrail.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
+"""
+aws_cloudtrail.py
+
+An ansible-rulebook event source module for getting events from an AWS CloudTrail
+
+Arguments:
+    connection - Parameters used to create AWS session
+    lookup_attributes - A dictionary representing the filters to be applied.
```
[the remark this answers is not in the document]
Comment (abikouo) [reply plugins/event_source/aws_cloudtrail.py:8]: updated

File: plugins/event_source/aws_cloudtrail.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+def _parse_user_inputs(**kwargs):
+    params = {}
+    attributes = kwargs.get("lookup_attributes")
+    if attributes:
+        filters_list = []
+        for k, v in attributes.items():
+            filter_dict = {"AttributeKey": k, "AttributeValue": v}
+            if isinstance(v, bool):
+                filter_dict["AttributeValue"] = str(v).lower()
+            elif isinstance(v, int):
```
[the remark this answers is not in the document]
Comment (abikouo) [reply plugins/event_source/aws_cloudtrail.py:44]: I will revert this and keep filtering as specified in the API

File: plugins/event_source/aws_cloudtrail.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
+"""
+aws_cloudtrail.py
+
+An ansible-rulebook event source module for getting events from an AWS CloudTrail
+
+Arguments:
+    connection        - Parameters used to create AWS session
```
Comment (bzwei) [inline plugins/event_source/aws_cloudtrail.py:7]: can you added supported parameters? https://aiobotocore.readthedocs.io/en/latest/tutorial.html

File: plugins/event_source/aws_cloudtrail.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                params["StartTime"] = event_time
+
+            events, c_event_time, c_event_ids = get_events(events, event_ids)
+            for event in events:
+                await queue.put(_cloudtrail_event_to_dict(event))
+
+            event_ids = c_event_ids or event_ids
+            event_time = c_event_time or event_time
+
+            await asyncio.sleep(delay)
```
Comment (bzwei) [inline plugins/event_source/aws_cloudtrail.py:100]: Is https://github.com/ansible/event-driven-ansible/blob/main/plugins/event_source/sqs.py#L46 only for sqs service or for all aws services? This way we don't need to do polling.

File: requirements.txt [hunk 8]
```diff
@@ -3,4 +3,6 @@ aiohttp
 aiokafka
 watchdog
 systemd-python
-dpath
\ No newline at end of file
+dpath
+botocore
```
[the remark this answers is not in the document]
Comment (abikouo) [reply requirements.txt:7]: yes sure

File: plugins/event_source/aws_cloudtrail.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+    - ansible.eda.aws_cloudtrail:
+        connection:
+            region_name: us-east-1
+        lookup_attributes:
+            - AttributeKey: 'EventSource'
+              AttributeValue: 'ec2.amazonaws.com'
+            - AttributeKey: 'ReadOnly'
+              AttributeValue: 'true'
+            Username: ansible
```
[the remark this answers is not in the document]
Comment (abikouo) [reply plugins/event_source/aws_cloudtrail.py:26]: mistake

File: plugins/event_source/aws_cloudtrail.py (hunk 7 above)
Comment (abikouo) [reply plugins/event_source/aws_cloudtrail.py:100]: I am not aware of this for the cloudtrail event, I need to investigate if we can use something similar and remove user polling

**Comment by abikouo** [comment]
@jillr @bzwei do you know who can merge this?

**Comment by bzwei** [comment]
> @jillr @bzwei do you know who can merge this?

Need one more approval

**Comment by abikouo** [comment]
@jillr @benthomasson could you please review/approve once again? Thanks

**Review by benthomasson — APPROVED** [review]

**Pull request merged**